diff --git a/ranger/container/fsobject.py b/ranger/container/fsobject.py
index 7de889bf..5ca9bf71 100644
--- a/ranger/container/fsobject.py
+++ b/ranger/container/fsobject.py
@@ -3,6 +3,7 @@
 
 from __future__ import (absolute_import, division, print_function)
 
+import mimetypes
 import re
 from grp import getgrgid
 from os import lstat, stat
@@ -216,7 +217,7 @@ class FileSystemObject(  # pylint: disable=too-many-instance-attributes,too-many
         if self.extension == 'part':
             bname = bname[0:-5]
         # pylint: disable=attribute-defined-outside-init
-        self._mimetype = self.fm.mimetypes.guess_type(bname, False)[0]
+        self._mimetype = mimetypes.guess_type(bname, False)[0]
         if self._mimetype is None:
             self._mimetype = ''
         # pylint: enable=attribute-defined-outside-init
diff --git a/ranger/core/fm.py b/ranger/core/fm.py
index 7d23c9b6..79b67d43 100644
--- a/ranger/core/fm.py
+++ b/ranger/core/fm.py
@@ -76,7 +76,6 @@ class FM(Actions,  # pylint: disable=too-many-instance-attributes
 
         mimetypes.knownfiles.append(os.path.expanduser('~/.mime.types'))
         mimetypes.knownfiles.append(self.relpath('data/mime.types'))
-        self.mimetypes = mimetypes.MimeTypes()
 
     def initialize(self):
         """If ui/bookmarks are None, they will be initialized here."""
diff --git a/ranger/core/fm.py b/ranger/core/fm.py
index 5b22164..a4fd2c7 100644
--- a/ranger/core/fm.py
+++ b/ranger/core/fm.py
@@ -151,7 +151,7 @@ class FM(Actions,  # pylint: disable=too-many-instance-attributes
 
             if self.settings.open_all_images and \
                     not self.thisdir.marked_items and \
-                    re.match(r'^(feh|sxiv|imv|pqiv) ', command):
+                    re.match(r'^(feh|nsxiv|sxiv|imv|pqiv) ', command):
 
                 images = [f.relative_path for f in self.thisdir.files if f.image]
                 escaped_filenames = " ".join(shell_quote(f) for f in images if "\x00" not in f)
@@ -160,7 +160,7 @@ class FM(Actions,  # pylint: disable=too-many-instance-attributes
                         "$@" in command:
                     new_command = None
 
-                    if command[0:5] == 'sxiv ':
+                    if command[0:5] == 'sxiv ' or command[0:6] == 'nsxiv ':
                         number = images.index(self.thisfile.relative_path) + 1
                         new_command = command.replace("sxiv ", "sxiv -n %d " % number, 1)
 
